[mlir][dataflow] disallow outside use of propagateIfChanged for DataFlowSolver

ZenithalHourlyRate opened the pull request:
Detailed writeup is in https://github.com/google/heir/issues/1153. See also https://github.com/llvm/llvm-project/pull/120881. In short, `propagateIfChanged` is used outside of the `DataFlowAnalysis` scope, because it is public, but it does not propagate as expected as the `DataFlowSolver` has stopped running.

To solve such misuse, `propagateIfChanged` should be made protected/private.

For downstream users affected by this, to correctly propagate the change, the Analysis should be re-run (check #120881) instead of just a `propagateIfChanged`

The change to `IntegerRangeAnalysis` is just a expansion of the `solver->propagateIfChanged`. The `Lattice` has already been updated by the `join`. Propagation is done by `onUpdate`.

Cc @Mogball for review

ZenithalHourlyRate committed:
[mlir][dataflow] Only allow propagateIfChanged when DataFlowSolver is running

Mogball commented on mlir/lib/Analysis/DataFlow/IntegerRangeAnalysis.cpp:
```diff
@@ -45,9 +45,13 @@ void IntegerValueRangeLattice::onUpdate(DataFlowSolver *solver) const {
-  if (!constant)
-    return solver->propagateIfChanged(
-        cv, cv->join(ConstantValue::getUnknownConstant()));
+  if (!constant) {
```
This is actually the intended usage of the `propagateIfChanged` method. Data flow variables that depend on each other can issue updates to each other through this method.

Mogball reviewed (changes requested):
Thanks for the PR and I agree that this being a public method is a bit of a footgun. However, directly invoking this method inside other dataflow variables is the intended usage of the method. Do you have a better idea on how an API can be exposed that way? `propagateIfChanged` is supposed to be a shorthand so that users don't forget to check the result of `join`, for example. Maybe it should be marked as `nodiscard`? (Unless it already is...). Alternatively, placing an assert inside `propagateIfChanged` checking that the solver is running might do the trick.

ZenithalHourlyRate commented:
> Alternatively, placing an assert inside propagateIfChanged checking that the solver is running might do the trick.

Added a `isRunning` flag. It is working but management of that flag should be careful (unset it in each return).

> Do you have a better idea on how an API can be exposed that way? 

I was thinking that adding `propagateIfChanged` to `AnalysisState` so all children `AnalysisState` could access it.  

```cpp

class DataFlowSolver {
  friend class AnalysisState;
};

class AnalysisState {

  // existing API
  virtual void onUpdate(DataFlowSolver *solver) const;

  // NOTE: it is not virtual
  void propagateIfChanged(DataFlowSolver *solver, AnalysisState *state, ChangeResult changed) {
    solver->propagateIfChanged(state, changed);
  }

};

```

Mogball commented on mlir/include/mlir/Analysis/DataFlowFramework.h:
```diff
@@ -412,6 +414,9 @@ class DataFlowSolver {
   /// Configuration of the dataflow solver.
   DataFlowConfig config;
 
+  // The solver is working on the worklist.
```
```suggestion
  /// The solver is working on the worklist.
```

Mogball commented on mlir/lib/Analysis/DataFlowFramework.cpp:
```diff
@@ -104,12 +104,16 @@ Location LatticeAnchor::getLoc() const {
 
 LogicalResult DataFlowSolver::initializeAndRun(Operation *top) {
+  // Enable enqueue to the worklist.
+  isRunning = true;
```
You can use `llvm::make_scope_exit` to ensure the flag is reset upon leaving the function

Mogball reviewed (commented):
Thanks for the ping. I think this looks alright. Just one nit

ZenithalHourlyRate commented:
Comments addressed.
